content/issue-2/optimize-binary-search.md

```diff
@@ -0,0 +1,509 @@
+
+# Branch prediction
+
+[Branch prediction] is a technique used by modern CPUs to predict the next possible branch in advance when encountering a branch in order to speed up the parallelization of instructions. A dedicated branch predictor is typically built into the CPU to support this functionality.
+
+For more information on branch prediction and how it affects performance, we recommend reading Stackoverflow's answer to ["Why is processing a sorted array faster than processing an unsorted array?"](https://stackoverflow.com/questions/11227809/why-is-processing-a-sorted-array-faster-than-processing-an-unsorted-array) This answer provides a clear explanation of branch prediction and its impact on performance.
+
+## Branch
+
+Before delving into branch prediction, it's important to understand what a `branch` is. At the high-level programming language, branches are statements such as `if/else/else if`, `goto`, or `switch/match`. However, these statements are ultimately converted into `jump` instructions in assembly code. In x86 assembly, `jump` instructions start with `j` and include:
+
+| instruction | effect                                                           |
+| ----------- | ---------------------------------------------------------------- |
+| jmp         | Always jump                                                      |
+| je          | Jump if cmp is equal                                             |
+| jne         | Jump if cmp is not equal                                         |
+| jg          | Signed > (greater)                                               |
+| jge         | Signed >=                                                        |
+| jl          | Signed < (less than)                                             |
+| jle         | Signed <=                                                        |
+| ja          | Unsigned > (above)                                               |
+| jae         | Unsigned >=                                                      |
+| jb          | Unsigned < (below)                                               |
+| jbe         | Unsigned <=                                                      |
+| jecxz       | Jump if ecx is 0                                                 |
+| jc          | Jump if carry: used for unsigned overflow, or multiprecision add |
+| jo          | Jump if there was signed overflow                                |
+
+To illustrate, we can write a Rust program that uses `if/else` statements, and look at the corresponding assembly code. For example:
+
+```rust
+#![allow(unused_assignments)]
+pub fn main() {
+  let mut a: usize = std::env::args().nth(1).unwrap().parse().unwrap_or_defaul();
+  if a > 42 {
+    a = 1;
+  } else {
+    a = 0;
+  }
+}
+```
+
+Here's the relevant portion of the resulting assembly code:
```

Comment:
Should we add a small pointer on how do we get to this assembly.
e.g what command to run etc etc
[btw, excellent and informative post so-far :)]

Reply:
Good point, I'll add a Godbolt link.